FogCreek/Glitch-Community
# Fix bugs for when project actions show up on the Teams page

This PR fixes https://glitch.manuscript.com/f/cases/edit/3326586/teams-remove-project-appears-on-team-page-IF-you-re-a-member-of-the-project-AND-not-a-member-of-the-team AND https://glitch.manuscript.com/f/cases/3326590/team-join-project-option-exposed-to-non-team-member-on-page

The root issue was that the teams page wasn’t looking at whether the user was a member of the team when considering whether or not the user is allowed to do certain actions. 
1. For removing the project from the teams page, we were looking at currentUserIsOnProject - we can completely ignore that field here, and consider only whether the user is part of the team (especially as a user can join any project in a team anyway).
2. For joining the project, we weren’t making sure the user was a member of the team before displaying the button.
3. Found and fixed an additional bug where you couldn’t leave a project you were a member of on a teams page you didn’t belong to.

The current design pattern for this component tree is to determine whether the user has the required permissions high in the component tree, and pass the function down only if they do (e.g. for add/removePin). I’ve therefore fixed it this way.

However after talking to Greg, we think this part of the site could do with some comprehensive refactoring - the logic around whether to allow actions on projects is scattered at many levels throughout the component tree, and there’s a bunch of duplication (e.g. EntityPageProjects and EntityPagePinnedProjects are almost exactly the same components). Created https://glitch.manuscript.com/f/cases/3327672/Refactor-Teams-Users-pages-to-simplify-logic-of-when-Project-Actions-show-up to track refactoring this, which Greg and I will pair on at some point soon.

Test plan:
1. Confirm that I cannot remove a project from a teams page that I’m not a member of (used @spotify) ✅
    1. Confirmed also for projects that I’m a member of on teams that I’m not a member of (e.g. community on @Performance-Boost ✅
2. Confirm that I can remove a project from one of my teams’ pages, where I am not a member of said project ✅
3. Confirm that the actions on the users page have not changed, for my projects that are also part of teams and also those that are not (as the code that was altered is shared between both teams and users profile pages) ✅

1. Confirm that I can’t join a project on a teams page that I’m not a member of ✅
2. Confirm that I can join a project on a teams page that I am a member of ✅
3. Confirm that I can’t join a project on a teams page that I am a member of, if I’m already a member of the project. ✅

1. Confirm that I can leave a project that I’m a member of on a team that I’m not a member of ✅
2. Confirm that I can’t leave a project that I’m not a member of ✅
